proposals/testing/NNNN-polling-confirmations.md

```diff
@@ -0,0 +1,668 @@
+    let subject = Aquarium()
+    subject.hasFunding = true
+
+    subject.raiseDolphins()
+
+    try await confirmation(until: .firstPass) { subject.isRaising == true }
+}
+
+@Test func `raiseDolphins if no funding keeps isRaising false`() async throws {
+    let subject = Aquarium()
+    subject.hasFunding = false
+
+    subject.raiseDolphins()
+
+    try await confirmation(until: .stopsPassing) { subject.isRaising == false }
+}
+```
+
+## Detailed design
+
+### New confirmation functions
+
+We will introduce 2 new members of the confirmation family of functions to the
+testing library:
+
+```swift
+/// Poll expression within the duration based on the given stop condition
+///
+/// - Parameters:
+///   - comment: A user-specified comment describing this confirmation.
+///   - stopCondition: When to stop polling.
+///   - duration: The expected length of time to continue polling for.
+///     This value does not incorporate the time to run `body`, and may not
+///     correspond to the wall-clock time that polling lasts for, especially on
+///     highly-loaded systems with a lot of tests running.
+///     If nil, this uses whatever value is specified under the last
+///     ``PollingConfirmationConfigurationTrait`` added to the test or suite
+///     with a matching stopCondition.
+///     If no such trait has been added, then polling will be attempted for
+///     about 1 second before recording an issue.
+///     `duration` must be greater than or equal to `interval`.
+///   - interval: The minimum amount of time to wait between polling attempts.
+///     If nil, this uses whatever value is specified under the last
+///     ``PollingConfirmationConfigurationTrait`` added to the test or suite
+///     with a matching stopCondition.
+///     If no such trait has been added, then polling will wait at least
+///     1 millisecond between polling attempts.
+///     `interval` must be greater than 0.
+///   - isolation: The actor to which `body` is isolated, if any.
+///   - sourceLocation: The location in source where the confirmation was called.
+///   - body: The function to invoke. The expression is considered to pass if
+///     the `body` returns true. Similarly, the expression is considered to fail
+///     if `body` returns false.
+///
+/// - Throws: A `PollingFailedError` if the `body` does not return true within
+///   the polling duration.
+///
+/// Use polling confirmations to check that an event while a test is running in
+/// complex scenarios where other forms of confirmation are insufficient. For
+/// example, waiting on some state to change that cannot be easily confirmed
+/// through other forms of `confirmation`.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public func confirmation(
+  _ comment: Comment? = nil,
+  until stopCondition: PollingStopCondition,
+  within duration: Duration? = nil,
+  pollingEvery interval: Duration? = nil,
+  isolation: isolated (any Actor)? = #isolation,
```

Comment:
Prefer expressing this by making the function “nonisolated(nonsending)”

Reply:
Thanks! I clearly missed the SE proposal that changed this.